File: EllipticalPath.cpp
#include "stdafx.h"
#include "EllipticalPath.h"
#include <math.h>


EllipticalPath::EllipticalPath(HWND hWnd, int window_width, int window_height)
{
	hWindow = hWnd; // save away window handle
	this->window_width = window_width; // save away window width
	this->window_height = window_height; // save away window width

	ball_colour = RGB(255, 0, 0);
	//ellipse coordinates
	h = 250;
	k = 275;
	a = 150;
	b = 100;
	x = h - a;
	y = k;

	// Pen and brush for travelling ball
	ball_pen = CreatePen(PS_SOLID, 1, RGB(0, 0, 0));
	ball_brush = CreateSolidBrush(ball_colour);

	hdc = GetDC(hWindow);

	SelectObject(hdc, ball_pen); // select ball pen colour
	SelectObject(hdc, ball_brush); // select ball brush colour
}

/*
* draws the ball/circle using the apt color
*/
void EllipticalPath::paint() {
	// draw a dot
	Ellipse(hdc, x, y, x + aWIDTH, y + aHEIGHT);
}

/*
* Repeatedly draws ball so as to simulate a continuous motion
*/
void EllipticalPath::moveElliptic() {
	// condition for continuing motion
	while (x <= h + a) {
		y = (int)round(k - ((double)b / a) * sqrt(pow(a, 2) - pow((x - h), 2)));
		paint();

		y = (int)round(k + ((double)b / a) * sqrt(pow(a, 2) - pow((x - h), 2)));
		paint();

		x += 20;
		// introduce a delay betweeen renderings
		Sleep(50);
	}
}


EllipticalPath::~EllipticalPath()
{
	DeleteObject(ball_pen);
	DeleteObject(ball_brush);

	ReleaseDC(hWindow, hdc);
}

File: EllipticalPath.java
package dymetric;

import java.awt.*;

public class EllipticalPath extends Canvas {

    protected Color ball_colour;
    //ellipse coordinates
    protected int h = 250;
    protected int k = 175;
    protected int a = 150;
    protected int b = 100;
    protected int x = h - a;
    protected int y = k;
    protected final int aWIDTH, aHEIGHT;

    public EllipticalPath() {
        setBackground(Color.LIGHT_GRAY); // canvas color
        ball_colour = Color.RED;
        aWIDTH = aHEIGHT = 10;
    }

    // Feel free to double buffer if flickering occurs
    public void paint(Graphics g) {
        g.setColor(ball_colour);
        // draw a dot
        g.fillOval(x, y, aWIDTH, aHEIGHT);
    }

    public void moveElliptic() {
        // condition for continuing motion
        while (x <= h + a) {
            y = (int)Math.round(k - ((double)b / a) * Math.sqrt(Math.pow(a, 2) - Math.pow((x - h), 2)));
            paint(this.getGraphics());

            y = (int)Math.round(k + ((double)b / a) * Math.sqrt(Math.pow(a, 2) - Math.pow((x - h), 2)));
            paint(this.getGraphics());
            
            x += 20;
            // introduce a delay betweeen renderings
            try {
                Thread.sleep(50);
            } catch (InterruptedException e) {
            }
        }
    }
}
